### PR TITLE
WPL

### PR DESCRIPTION
**Pull request by ROBOTICSENGINEER**
Weibull Prototype Learning (WPL)

### CODE REVIEW

File: vast/opensetAlgos/WPL.py [hunk 1]
```diff
@@ -0,0 +1,140 @@
+
+"""
+Author: Akshay Raj Dhamija
+
+Weibull Prototype Learning (WPL)
```
[the remark this answers is not in the document]
Comment (akshay-raj-dhamija) [reply vast/opensetAlgos/WPL.py:5]: Thanks!

File: vast/opensetAlgos/WPL.py [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+        default=[1.0],
+        help="Weibull scale to use when the number of uniqe element is less than 5 default: %(default)s",
+    )
+    WPL_params.add_argument(
+        "--default_shape",
+        nargs="+",
+        type=float,
+        default=[0.1],
+        help="Weibull shape to use when the number of uniqe element is less than 5 default:  %(default)s",
+    )
```
[the remark this answers is not in the document]
Comment (akshay-raj-dhamija) [reply vast/opensetAlgos/WPL.py:67]: **Algo Breakdown:** Please try to simplify or break down your algorithm. Your WPL algo has two parts one is per dimension weibull and the other is how you address the case where weibull is needed for less than 5 samples/values. I am asking to separate this part from your core algorithm.
If tomorrow someone comes up with a new way to estimate `default_scale` and `default_shape` rather than fixing default values (let's say a mean from other class values, which can't be found until the algorithm has been run on the entire dataset) would you create a new `WPL.py` and call it a new algorithm even though the per dimension weibull idea is the same?

**Reducing Computation:** `default_scale` and `default_shape` are not similar to `tail_size, cover_threshold, and distance_multiplier` of EVM. These EVM parameters were being used as a list to reduce redundant computation. In your case, you will be recalling the `mr.FitHigh` function using the same parameters when your `default_scale` and `default_shape` change, even though they do not impact the result in this case. That is you are adding computational overhead rather than reducing it.
`default_scale` and `default_shape` are simply being used to replace scale and shape values where none could be found, not reduce computation.

File: vast/opensetAlgos/WPL.py [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+    return parser, dict(
+        group_parser=WPL_params,
+        param_names=("tailsize", "distance_multiplier", "default_scale", "default_shape"),
+        param_id_string="TS_{}_DM_{:.4f}_SC_{:.4f}_SH_{:.4f}",
+    )
+
+
+def fit_high(distances, distance_multiplier, tailsize, default_shape, default_scale):
+    distances = torch.unique(distances)
```
[the remark this answers is not in the document]
Comment (akshay-raj-dhamija) [reply vast/opensetAlgos/WPL.py:76]: Even in openmax `no_of_samples_to_which_distance_was_found=1` but still its `fit_high` function supports working on a 2D tensor absence of `dim` in your code above removes that support.

I will let you and @tboult decide if you want to incorporate this in `weibull.py` it is very much possible to incorporate this there without changing the default behaviour for any of the other algorithms. At least changing it there helps to quickly test the same idea for other algorithms as well.

File: vast/opensetAlgos/WPL.py [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+    if tailsize <= 1:
+        tailsize = min(tailsize * distances.shape[1], distances.shape[1])
+    tailsize = int(min(tailsize, distances.shape[1]))
+    mr = weibull.weibull()
+    if distances.shape[1] < 5:
+        pass
```
[the remark this answers is not in the document]
Comment (akshay-raj-dhamija) [reply vast/opensetAlgos/WPL.py:82]: Please check my comment above https://github.com/Vastlab/vast/pull/37#discussion_r739441786

File: vast/opensetAlgos/WPL.py [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+    :param models: Not used during training, input ignored.
+    :return: Iterator(Tuple(parameter combination identifier, Tuple(class name, its evm model)))
+    """
+    with torch.no_grad():
+        for pos_cls_name in pos_classes_to_process:
+            features = features_all_classes[pos_cls_name].clone().to(f"cuda:{gpu}")
+            assert args.dimension == features.shape[1]
+
+            center = torch.mean(features, dim=0).to(f"cuda:{gpu}")
+            distances = torch.abs(features - center.view(1,args.dimension).repeat(features.shape[0], 1))
```
[the remark this answers is not in the document]
Comment (akshay-raj-dhamija) [reply vast/opensetAlgos/WPL.py]: yes, you are right that this might not be replaceable by other metrics and neither would be a useful metric for other algorithms, which is what I was thinking.
BTW, you do not need `.repeat` above, even if needed it should be replaced with `expand` as it reduces memory consumption.

File: vast/opensetAlgos/WPL.py [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+            for tailsize, distance_multiplier, default_shape, default_scale  in itertools.product(
+                args.tailsize, args.distance_multiplier, args.default_shape, args.default_scale
+            ):
+                  weibull_list = list()
+                  for k in args.dimension:
```
[the remark this answers is not in the document]
Comment (akshay-raj-dhamija) [reply vast/opensetAlgos/WPL.py]: You should also be able to use it to return multiple weibulls at once. If `distances` is a `nxk` tensor, where in your case `n` is the number of dimensions, then it can straightaway provide you `n` weibulls in a single mr object which will reduce both training and inference time.

File: vast/opensetAlgos/WPL.py [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+    for batch_to_process in pos_classes_to_process:
+        test_cls_feature = features_all_classes[batch_to_process].to(f"cuda:{gpu}")
+        assert test_cls_feature.shape[0] != 0
+        probs = []
+        for cls_no, cls_name in enumerate(models.keys())     
+            center = model[cls_name]["center"].to(f"cuda:{gpu}")
+            distances = torch.abs(test_cls_feature - center.view(1,args.dimension).repeat(test_cls_feature.shape[0], 1))
+            weibull_list = models[class_name]["weibull_list"]
+            p = torch.empty(args.dimension)
+            for k in args.dimension:
```
[the remark this answers is not in the document]
Comment (akshay-raj-dhamija) [reply vast/opensetAlgos/WPL.py]: Check comment above. https://github.com/Vastlab/vast/pull/37#discussion_r739443911

File: vast/opensetAlgos/WPL.py [hunk 8]
```diff
@@ -0,0 +1,172 @@
...
+    tailsize = int(min(tailsize, distances.shape[1]))
+    mr = weibull.weibull()
+    if distances.shape[1] < 5:
+        pass
+        mr.sign = 1
+        mr.wbFits = torch.zeros(1, 2)
+        mr.wbFits[0, 1] = default_scale
+        mr.wbFits[0, 0] = default_shape
+        mr._ = torch.Tensor(0.0) # translate Amount
+        mr.smallScoreTensor =  torch.Tensor(0.0)#  small Score
```
Comment (akshay-raj-dhamija) [inline vast/opensetAlgos/WPL.py:88]: You could just create this as done at https://github.com/Vastlab/vast/blob/460480fc3bf624d1ffee70be5ec446bf6f6ceee6/vast/opensetAlgos/multimodal_openmax.py#L119

File: vast/opensetAlgos/WPL.py [hunk 9]
```diff
@@ -0,0 +1,172 @@
...
+    :param models: Not used during training, input ignored.
+    :return: Iterator(Tuple(parameter combination identifier, Tuple(class name, its evm model)))
+    """
+    dimension = None
+    for pos_cls_name in pos_classes_to_process:
+        features = features_all_classes[pos_cls_name].clone().to(f"cuda:{gpu}")
+        if dimension == None:
+            dimension = features.shape[1]
+         else:
+            assert dimension == features.shape[1]
```
Comment (akshay-raj-dhamija) [inline vast/opensetAlgos/WPL.py:116]: `dimension` will always be `None` so will never reach the else statement.

File: vast/opensetAlgos/WPL.py [hunk 10]
```diff
@@ -0,0 +1,172 @@
...
+                function above. Only the distance_metric argument is actually used during inferencing.
+    :param gpu: An integer corresponding to the gpu number to use by the current process.
+    :param models: The collated model created for a single hyper parameter combination.
+    :return: Iterator(Tuple(str, Tuple(batch_identifier, torch.Tensor)))
+    """         
+    for batch_to_process in pos_classes_to_process:
+        test_cls_feature = features_all_classes[batch_to_process].to(f"cuda:{gpu}")
+        assert test_cls_feature.shape[0] != 0
+        probs = []
+        for cls_no, cls_name in enumerate(models.keys())
```
Comment (akshay-raj-dhamija) [inline vast/opensetAlgos/WPL.py:157]: While you expect the `models` variable to be an ordered dict if a user instead passes a dict rather than letting the user know this will simply provide them with incorrect results. This can be avoided by simply using  `enumerate(sorted(models.keys()))` instead of `enumerate(models.keys())`

File: vast/opensetAlgos/WPL.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (akshay-raj-dhamija) [reply vast/opensetAlgos/WPL.py:76]: _Correction:_ I made an incorrect suggestion above, `dim` with unique doesn't do what I expected it to. So using `unique` in `fit_high` will by default remove the support for computation on a 2D tensor.

File: vast/opensetAlgos/WPL.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (ROBOTICSENGINEER) [reply vast/opensetAlgos/WPL.py:67]: Both of the comments `Algo Breakdown` and `Reducing Computation` are correct. @scruz13 please make required changes.
